[TEST-FAILURE] executor.CompletableFutureTest.postregisterCallback

```
java.lang.AssertionError: CountDownLatch failed to complete within 120 seconds , count left: 1
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:718)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:711)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:703)
	at com.hazelcast.executor.CompletableFutureTest.postregisterCallback(CompletableFutureTest.java:104)
```

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x/com.hazelcast$hazelcast/3630/testReport/junit/com.hazelcast.executor/CompletableFutureTest/postregisterCallback/

Managed to reproduce locally. With ```@Repeat``` that's big enough it eventually fails

```
java.lang.AssertionError: CountDownLatch failed to complete within 120 seconds , count left: 1
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:717)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:710)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:702)
	at com.hazelcast.executor.CompletableFutureTest.postregisterCallback(CompletableFutureTest.java:106)
	at sun.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at com.hazelcast.test.AbstractHazelcastClassRunner$ThreadDumpAwareRunAfters.evaluate(AbstractHazelcastClassRunner.java:134)
	at com.hazelcast.test.AbstractHazelcastClassRunner$TestRepeater.evaluate(AbstractHazelcastClassRunner.java:219)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at com.hazelcast.test.HazelcastParallelClassRunner.access$201(HazelcastParallelClassRunner.java:35)
	at com.hazelcast.test.HazelcastParallelClassRunner$TestRunner.run(HazelcastParallelClassRunner.java:119)
	at java.lang.Thread.run(Thread.java:745)
```

--- a/hazelcast/src/main/java/com/hazelcast/mapreduce/impl/task/TrackableJobFuture.java
+++ b/hazelcast/src/main/java/com/hazelcast/mapreduce/impl/task/TrackableJobFuture.java
@@ -32,12 +32,7 @@ import com.hazelcast.spi.impl.NodeEngineImpl;
 import java.util.Arrays;
 import java.util.Map;
 import java.util.concurrent.CancellationException;
-import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.ExecutionException;
-import java.util.concurrent.TimeUnit;
-import java.util.concurrent.TimeoutException;
-
-import static com.hazelcast.util.Preconditions.isNotNull;
 
 /**
  * This is the node based implementation of the job's reactive {@link com.hazelcast.core.ICompletableFuture}
@@ -55,7 +50,6 @@ public class TrackableJobFuture<V>
     private final String jobId;
     private final JobTracker jobTracker;
     private final Collator collator;
-    private final CountDownLatch latch;
     private final MapReduceService mapReduceService;
 
     private volatile boolean cancelled;
@@ -66,34 +60,29 @@ public class TrackableJobFuture<V>
         this.jobId = jobId;
         this.jobTracker = jobTracker;
         this.collator = collator;
-        this.latch = new CountDownLatch(1);
         this.mapReduceService = ((NodeEngineImpl) nodeEngine).getService(MapReduceService.SERVICE_NAME);
     }
 
     @Override
     public void setResult(Object result) {
-        try {
-            Object finalResult = result;
-            if (finalResult instanceof Throwable && !(finalResult instanceof CancellationException)) {
-                super.setResult(new ExecutionException((Throwable) finalResult));
-                return;
-            }
-            // If collator is available we need to execute it now
-            if (collator != null) {
-                try {
-                    finalResult = collator.collate(((Map) finalResult).entrySet());
-                } catch (Exception e) {
-                    // Possible exception while collating
-                    finalResult = e;
-                }
-            }
-            if (finalResult instanceof Throwable && !(finalResult instanceof CancellationException)) {
-                finalResult = new ExecutionException((Throwable) finalResult);
-            }
-            super.setResult(finalResult);
-        } finally {
-            latch.countDown();
+        Object finalResult = result;
+        if (finalResult instanceof Throwable && !(finalResult instanceof CancellationException)) {
+            super.setResult(new ExecutionException((Throwable) finalResult));
+            return;
         }
+        // If collator is available we need to execute it now
+        if (collator != null) {
+            try {
+                finalResult = collator.collate(((Map) finalResult).entrySet());
+            } catch (Exception e) {
+                // Possible exception while collating
+                finalResult = e;
+            }
+        }
+        if (finalResult instanceof Throwable && !(finalResult instanceof CancellationException)) {
+            finalResult = new ExecutionException((Throwable) finalResult);
+        }
+        super.setResult(finalResult);
     }
 
     @Override
@@ -107,23 +96,8 @@ public class TrackableJobFuture<V>
             return false;
         }
         Exception exception = new CancellationException("Operation was cancelled by the user");
-        cancelled = supervisor.cancelAndNotify(exception);
+        cancelled = supervisor.cancelAndNotify(exception) && super.cancel(mayInterruptIfRunning);
         return cancelled;
-    }
-
-    @Override
-    public boolean isCancelled() {
-        return cancelled;
-    }
-
-    @Override
-    public V get(long timeout, TimeUnit unit)
-            throws InterruptedException, ExecutionException, TimeoutException {
-        isNotNull(unit, "unit");
-        if (!latch.await(timeout, unit) || !isDone()) {
-            throw new TimeoutException("timeout reached");
-        }
-        return getResult();
     }
 
     @Override

--- a/hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractCompletableFuture.java
+++ b/hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractCompletableFuture.java
@@ -22,6 +22,7 @@ import com.hazelcast.logging.ILogger;
 import com.hazelcast.spi.ExecutionService;
 import com.hazelcast.spi.NodeEngine;
 
+import java.util.concurrent.CancellationException;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Executor;
 import java.util.concurrent.TimeUnit;
@@ -32,9 +33,17 @@ import static com.hazelcast.util.ExceptionUtil.sneakyThrow;
 import static com.hazelcast.util.Preconditions.isNotNull;
 import static java.util.concurrent.atomic.AtomicReferenceFieldUpdater.newUpdater;
 
+/**
+ * A base {@link ICompletableFuture} implementation that may be explicitly completed by setting its
+ * value through setResult.
+ * Implements the logic of cancellation and callbacks execution.
+ *
+ * @param <V> The result type returned by this Future's {@code get} method
+ */
 public abstract class AbstractCompletableFuture<V> implements ICompletableFuture<V> {
 
-    static final Object INITIAL_STATE = new ExecutionCallbackNode(null, null, null);
+    private static final Object INITIAL_STATE = new ExecutionCallbackNode(null, null, null);
+    private static final Object CANCELLED_STATE = new Object();
 
     private static final AtomicReferenceFieldUpdater<AbstractCompletableFuture, Object> STATE
             = newUpdater(AbstractCompletableFuture.class, Object.class, "state");
@@ -45,7 +54,7 @@ public abstract class AbstractCompletableFuture<V> implements ICompletableFuture
     // The reason for this abuse of the type system is to deal with the head node and the result
     // in a single cas. Using a single cas prevent a thread that calls and then do this concurrently
     // with a thread that calls setResult.
-    volatile Object state = INITIAL_STATE;
+    private volatile Object state = INITIAL_STATE;
 
     private final ILogger logger;
     private final Executor defaultExecutor;
@@ -72,7 +81,11 @@ public abstract class AbstractCompletableFuture<V> implements ICompletableFuture
         for (; ; ) {
             Object currentState = this.state;
 
-            if (isDone(currentState)) {
+            if (isCancelledState(currentState)) {
+                return;
+            }
+
+            if (isDoneState(currentState)) {
                 runAsynchronous(callback, executor, currentState);
                 return;
             }
@@ -93,11 +106,58 @@ public abstract class AbstractCompletableFuture<V> implements ICompletableFuture
 
     @Override
     public boolean isDone() {
-        return isDone(state);
+        return isDoneState(state);
     }
 
-    private boolean isDone(Object state) {
+    /**
+     * Returns {@code true} if the task with the given state completed - analogously to the Future's contract.
+     * Completion may be due to normal termination, an exception, or
+     * cancellation -- in all of these cases, this method will return
+     * {@code true}.
+     *
+     * @return {@code true} if this task completed
+     */
+    private static boolean isDoneState(Object state) {
         return !(state instanceof ExecutionCallbackNode);
+    }
+
+    @Override
+    public boolean cancel(boolean mayInterruptIfRunning) {
+        for (; ; ) {
+            Object currentState = this.state;
+
+            if (isDoneState(currentState)) {
+                return false;
+            }
+
+            if (STATE.compareAndSet(this, currentState, CANCELLED_STATE)) {
+                cancelled(mayInterruptIfRunning);
+                notifyThreadsWaitingOnGet();
+                return true;
+            }
+        }
+    }
+
+    /**
+     * Protected method invoked when this task transitions to state
+     * {@code isCancelled}. The default implementation does nothing.
+     * Subclasses may override this method to invoke callbacks or perform
+     * bookkeeping. Implementation has to handle exceptions itself.
+     *
+     * @param mayInterruptIfRunning {@code true} if the thread executing this
+     *                              task was supposed to be interrupted; otherwise, in-progress tasks are allowed
+     *                              to complete
+     */
+    protected void cancelled(boolean mayInterruptIfRunning) {
+    }
+
+    private static boolean isCancelledState(Object state) {
+        return state == CANCELLED_STATE;
+    }
+
+    @Override
+    public boolean isCancelled() {
+        return isCancelledState(state);
     }
 
     @Override
@@ -110,25 +170,71 @@ public abstract class AbstractCompletableFuture<V> implements ICompletableFuture
         }
     }
 
+    @Override
+    public V get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
+        final long deadlineTimeMillis = System.currentTimeMillis() + unit.toMillis(timeout);
+        long millisToWait;
+        for (; ; ) {
+            Object currentState = this.state;
+
+            if (isCancelledState(currentState)) {
+                throw new CancellationException();
+            }
+            if (isDoneState(currentState)) {
+                return getResult();
+            }
+            if (Thread.interrupted()) {
+                throw new InterruptedException();
+            }
+
+            millisToWait = deadlineTimeMillis - System.currentTimeMillis();
+            if (millisToWait <= 0) {
+                throw new TimeoutException();
+            }
+
+            synchronized (this) {
+                if (!isDoneState(this.state)) {
+                    this.wait(millisToWait);
+                }
+            }
+        }
+    }
+
     public void setResult(Object result) {
         for (; ; ) {
             Object currentState = this.state;
 
-            if (isDone(currentState)) {
+            if (isDoneState(currentState)) {
                 return;
             }
 
             if (STATE.compareAndSet(this, currentState, result)) {
+                done();
+                notifyThreadsWaitingOnGet();
                 runAsynchronous((ExecutionCallbackNode) currentState, result);
                 break;
             }
         }
     }
 
+    /**
+     * Protected method invoked when this task transitions to state
+     * {@code isDone} (only normally - in case of cancellation cancelled() is invoked)
+     * The default implementation does nothing.
+     * Subclasses may override this method to invoke callbacks or perform
+     * bookkeeping. Implementation has to handle exceptions itself.
+     */
+    protected void done() {
+    }
+
     protected V getResult() {
         Object state = this.state;
 
-        if (!isDone(state)) {
+        if (isCancelledState(state)) {
+            return null;
+        }
+
+        if (!isDoneState(state)) {
             return null;
         }
 
@@ -137,6 +243,12 @@ public abstract class AbstractCompletableFuture<V> implements ICompletableFuture
         }
 
         return (V) state;
+    }
+
+    private void notifyThreadsWaitingOnGet() {
+        synchronized (this) {
+            this.notifyAll();
+        }
     }
 
     protected void runAsynchronous(ExecutionCallbackNode head, Object result) {

--- a/hazelcast/src/main/java/com/hazelcast/util/executor/CompletableFutureTask.java
+++ b/hazelcast/src/main/java/com/hazelcast/util/executor/CompletableFutureTask.java
@@ -16,123 +16,77 @@
 
 package com.hazelcast.util.executor;
 
-import com.hazelcast.core.ExecutionCallback;
 import com.hazelcast.core.ICompletableFuture;
-import com.hazelcast.logging.ILogger;
 import com.hazelcast.logging.Logger;
+import com.hazelcast.spi.impl.AbstractCompletableFuture;
 
 import java.util.concurrent.Callable;
-import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ExecutorService;
-import java.util.concurrent.FutureTask;
+import java.util.concurrent.Executors;
+import java.util.concurrent.RunnableFuture;
 import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
 
-import static com.hazelcast.util.Preconditions.isNotNull;
+import static java.util.concurrent.atomic.AtomicReferenceFieldUpdater.newUpdater;
 
-public class CompletableFutureTask<V> extends FutureTask<V> implements ICompletableFuture<V> {
+public class CompletableFutureTask<V> extends AbstractCompletableFuture<V> implements ICompletableFuture<V>, RunnableFuture<V> {
 
-    private final AtomicReferenceFieldUpdater<CompletableFutureTask, ExecutionCallbackNode> callbackUpdater;
-    private final ILogger logger = Logger.getLogger(CompletableFutureTask.class);
-    private final ExecutorService asyncExecutor;
+    private static final AtomicReferenceFieldUpdater<CompletableFutureTask, Thread> RUNNER
+            = newUpdater(CompletableFutureTask.class, Thread.class, "runner");
 
-    private volatile ExecutionCallbackNode<V> callbackHead;
+    private final Callable<V> callable;
+
+    private volatile Thread runner;
 
     public CompletableFutureTask(Callable<V> callable, ExecutorService asyncExecutor) {
-        super(callable);
-        this.asyncExecutor = asyncExecutor;
-        this.callbackUpdater = AtomicReferenceFieldUpdater.newUpdater(
-                CompletableFutureTask.class, ExecutionCallbackNode.class, "callbackHead");
+        super(asyncExecutor, Logger.getLogger(CompletableFutureTask.class));
+        this.callable = callable;
     }
 
     public CompletableFutureTask(Runnable runnable, V result, ExecutorService asyncExecutor) {
-        super(runnable, result);
-        this.asyncExecutor = asyncExecutor;
-        this.callbackUpdater = AtomicReferenceFieldUpdater.newUpdater(
-                CompletableFutureTask.class, ExecutionCallbackNode.class, "callbackHead");
+        super(asyncExecutor, Logger.getLogger(CompletableFutureTask.class));
+        this.callable = Executors.callable(runnable, result);
     }
 
     @Override
     public void run() {
-        try {
-            super.run();
-        } finally {
-            fireCallbacks();
-        }
-    }
-
-    @Override
-    public void andThen(ExecutionCallback<V> callback) {
-        andThen(callback, asyncExecutor);
-    }
-
-    @Override
-    public void andThen(ExecutionCallback<V> callback, Executor executor) {
-        isNotNull(callback, "callback");
-        isNotNull(executor, "executor");
-
         if (isDone()) {
-            runAsynchronous(callback, executor);
             return;
         }
-        for ( ;;) {
-            ExecutionCallbackNode<V> oldCallbackHead = callbackHead;
-            ExecutionCallbackNode<V> newCallbackHead = new ExecutionCallbackNode<V>(callback, executor, oldCallbackHead);
-            if (callbackUpdater.compareAndSet(this, oldCallbackHead, newCallbackHead)) {
-                break;
-            }
-        }
-    }
 
-    private Object readResult() {
+        if (runner != null || !RUNNER.compareAndSet(this, null, Thread.currentThread())) {
+            // prevents concurrent calls to run
+            return;
+        }
+
         try {
-            return get();
-        } catch (Throwable t) {
-            return t;
-        }
-    }
-
-    private void fireCallbacks() {
-        ExecutionCallbackNode<V> callbackChain;
-        for (;;) {
-            callbackChain = callbackHead;
-            if (callbackUpdater.compareAndSet(this, callbackChain, null)) {
-                break;
-            }
-        }
-        while (callbackChain != null) {
-            runAsynchronous(callbackChain.callback, callbackChain.executor);
-            callbackChain = callbackChain.next;
-        }
-    }
-
-    private void runAsynchronous(final ExecutionCallback<V> callback, final Executor executor) {
-        final Object result = readResult();
-        executor.execute(new Runnable() {
-            @Override
-            public void run() {
+            Callable c = callable;
+            if (c != null) {
+                Object result = null;
                 try {
-                    if (result instanceof Throwable) {
-                        callback.onFailure((Throwable) result);
-                    } else {
-                        callback.onResponse((V) result);
-                    }
-                } catch (Throwable t) {
-                    //todo: improved error message
-                    logger.severe("Failed to async for " + CompletableFutureTask.this, t);
+                    result = c.call();
+                } catch (Throwable ex) {
+                    result = new ExecutionException(ex);
+                } finally {
+                    setResult(result);
                 }
             }
-        });
-    }
-
-    private static final class ExecutionCallbackNode<E> {
-        private final ExecutionCallback<E> callback;
-        private final Executor executor;
-        private final ExecutionCallbackNode<E> next;
-
-        private ExecutionCallbackNode(ExecutionCallback<E> callback, Executor executor, ExecutionCallbackNode<E> next) {
-            this.callback = callback;
-            this.executor = executor;
-            this.next = next;
+        } finally {
+            // runner must be non-null until state is settled in setResult() to
+            // prevent concurrent calls to run()
+            runner = null;
         }
     }
+
+    @Override
+    protected void cancelled(boolean mayInterruptIfRunning) {
+        // additionally handle the interruption of the executing thread
+        if (mayInterruptIfRunning) {
+            Thread executingThread = runner;
+            if (executingThread != null) {
+                executingThread.interrupt();
+            }
+        }
+    }
+
 }

--- a/hazelcast/src/test/java/com/hazelcast/executor/CompletableFutureTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/executor/CompletableFutureTest.java
@@ -23,158 +23,350 @@ import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.HazelcastTestSupport;
 import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
+import com.hazelcast.test.annotation.Repeat;
 import org.hamcrest.Matchers;
-import org.junit.After;
 import org.junit.Before;
+import org.junit.Rule;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
+import org.junit.rules.ExpectedException;
 import org.junit.runner.RunWith;
 
 import java.util.concurrent.Callable;
+import java.util.concurrent.CancellationException;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
 import java.util.concurrent.atomic.AtomicReference;
 
 import static org.hamcrest.Matchers.instanceOf;
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
 
 @RunWith(HazelcastParallelClassRunner.class)
 @Category({QuickTest.class, ParallelTest.class})
 public class CompletableFutureTest extends HazelcastTestSupport {
 
-    private static final RuntimeException TEST_EXCEPTION = new RuntimeException("Test exception");
+    private static final RuntimeException THROW_TEST_EXCEPTION = new RuntimeException("Test exception");
+    private static final RuntimeException NO_EXCEPTION = null;
 
     private ExecutionService executionService;
-    private CountDownLatch startLatch, doneLatch;
-    private AtomicReference<Object> ref1, ref2;
+    private CountDownLatch inExecutionLatch, startLogicLatch, executedLogic, finishedLatch, callbacksDoneLatch;
+    private AtomicReference<Object> reference1, reference2;
+
+    @Rule
+    public ExpectedException expected = ExpectedException.none();
 
     @Before
     public void setUp() throws Exception {
         NodeEngine nodeEngine = getNode(createHazelcastInstance()).getNodeEngine();
         executionService = nodeEngine.getExecutionService();
-        startLatch = new CountDownLatch(1);
-        ref1 = new AtomicReference<Object>();
-        ref2 = new AtomicReference<Object>();
-    }
-
-    @After
-    public void tearDown() {
+        startLogicLatch = new CountDownLatch(1);
+        executedLogic = new CountDownLatch(1);
+        finishedLatch = new CountDownLatch(1);
+        inExecutionLatch = new CountDownLatch(1);
+        reference1 = new AtomicReference<Object>();
+        reference2 = new AtomicReference<Object>();
     }
 
     @Test
     public void preregisterCallback() throws Exception {
-        doneLatch = new CountDownLatch(1);
-        final ICompletableFuture<String> f = submit(awaitStartLatch());
-        f.andThen(setRefAndBumpDoneLatch(ref1));
-        startLatch.countDown();
-        assertOpenEventually(doneLatch);
-        assertEquals("success", ref1.get());
+        ICompletableFuture<String> f = submitAwaitingTask(expectedNumberOfCallbacks(1), NO_EXCEPTION);
+        f.andThen(storeTaskResponseToReference(reference1));
+
+        releaseAwaitingTask();
+
+        assertCallbacksExecutedEventually();
+        assertEquals("success", reference1.get());
     }
 
     @Test
     public void preregisterTwoCallbacks() throws Exception {
-        doneLatch = new CountDownLatch(2);
-        final ICompletableFuture<String> f = submit(awaitStartLatch());
-        f.andThen(setRefAndBumpDoneLatch(ref1));
-        f.andThen(setRefAndBumpDoneLatch(ref2));
-        startLatch.countDown();
-        assertOpenEventually(doneLatch);
-        assertEquals("success", ref1.get());
-        assertEquals("success", ref2.get());
+        ICompletableFuture<String> f = submitAwaitingTask(expectedNumberOfCallbacks(2), NO_EXCEPTION);
+        f.andThen(storeTaskResponseToReference(reference1));
+        f.andThen(storeTaskResponseToReference(reference2));
+
+        releaseAwaitingTask();
+
+        assertCallbacksExecutedEventually();
+        assertEquals("success", reference1.get());
+        assertEquals("success", reference2.get());
     }
 
     @Test
-    public void preregisterTwoCallbacks_withFailure() throws Exception {
-        doneLatch = new CountDownLatch(2);
-        final ICompletableFuture<String> f = submit(awaitStartLatch(), throwException());
-        f.andThen(setRefAndBumpDoneLatch(ref1));
-        f.andThen(setRefAndBumpDoneLatch(ref2));
-        startLatch.countDown();
-        assertOpenEventually(doneLatch);
-        assertTestException(ref1, ref2);
+    public void preregisterTwoCallbacks_taskThrowsException() throws Exception {
+        ICompletableFuture<String> f = submitAwaitingTask(expectedNumberOfCallbacks(2), THROW_TEST_EXCEPTION);
+        f.andThen(storeTaskResponseToReference(reference1));
+        f.andThen(storeTaskResponseToReference(reference2));
+
+        releaseAwaitingTask();
+
+        assertCallbacksExecutedEventually();
+        assertTestExceptionThrown(reference1, reference2);
     }
 
     @Test
+    @Repeat(10)
+    // https://github.com/hazelcast/hazelcast/issues/6020
     public void postregisterCallback() throws Exception {
-        doneLatch = new CountDownLatch(1);
-        final ICompletableFuture<String> f = submit(openStartLatch());
-        assertOpenEventually(startLatch);
-        f.andThen(setRefAndBumpDoneLatch(ref1));
-        assertOpenEventually(doneLatch);
-        assertEquals("success", ref1.get());
+        ICompletableFuture<String> f = submitAwaitingTask(expectedNumberOfCallbacks(1), NO_EXCEPTION);
+        releaseAwaitingTask();
+        assertTaskFinishedEventually();
+
+        f.andThen(storeTaskResponseToReference(reference1));
+
+        assertCallbacksExecutedEventually();
+        assertEquals("success", reference1.get());
     }
 
     @Test
     public void postregisterTwoCallbacks() throws Exception {
-        doneLatch = new CountDownLatch(2);
-        final ICompletableFuture<String> f = submit(openStartLatch());
-        assertOpenEventually(startLatch);
-        f.andThen(setRefAndBumpDoneLatch(ref1));
-        f.andThen(setRefAndBumpDoneLatch(ref2));
-        assertOpenEventually(doneLatch);
-        assertEquals("success", ref1.get());
-        assertEquals("success", ref2.get());
+        ICompletableFuture<String> f = submitAwaitingTask(expectedNumberOfCallbacks(2), NO_EXCEPTION);
+        releaseAwaitingTask();
+        assertTaskFinishedEventually();
+
+        f.andThen(storeTaskResponseToReference(reference1));
+        f.andThen(storeTaskResponseToReference(reference2));
+
+        assertCallbacksExecutedEventually();
+        assertEquals("success", reference1.get());
+        assertEquals("success", reference2.get());
     }
 
     @Test
-    public void postregisterTwoCallbacks_withFailure() throws Exception {
-        doneLatch = new CountDownLatch(2);
-        final ICompletableFuture<String> f = submit(openStartLatch(), throwException());
-        assertOpenEventually(startLatch);
-        f.andThen(setRefAndBumpDoneLatch(ref1));
-        f.andThen(setRefAndBumpDoneLatch(ref2));
-        assertOpenEventually(doneLatch);
-        assertTestException(ref1, ref2);
+    public void postregisterTwoCallbacks_taskThrowsException() throws Exception {
+        ICompletableFuture<String> f = submitAwaitingTask(expectedNumberOfCallbacks(2), THROW_TEST_EXCEPTION);
+        releaseAwaitingTask();
+        assertTaskFinishedEventually();
+
+        f.andThen(storeTaskResponseToReference(reference1));
+        f.andThen(storeTaskResponseToReference(reference2));
+
+        assertCallbacksExecutedEventually();
+        assertTestExceptionThrown(reference1, reference2);
     }
 
-    private static void assertTestException(AtomicReference<?>... refs) {
+    @Test(timeout = 60000)
+    public void get_taskThrowsException() throws Exception {
+        ICompletableFuture<String> f = submitAwaitingTaskNoCallbacks(THROW_TEST_EXCEPTION);
+        submitReleasingTask(100);
+
+        expected.expect(ExecutionException.class);
+        f.get();
+    }
+
+    @Test(timeout = 60000)
+    public void getWithTimeout_taskThrowsException() throws Exception {
+        ICompletableFuture<String> f = submitAwaitingTaskNoCallbacks(THROW_TEST_EXCEPTION);
+        submitReleasingTask(200);
+
+        expected.expect(ExecutionException.class);
+        f.get(30000, TimeUnit.MILLISECONDS);
+    }
+
+    @Test(timeout = 60000)
+    @Repeat
+    public void getWithTimeout_finishesWithinTime() throws Exception {
+        ICompletableFuture<String> f = submitAwaitingTaskNoCallbacks(NO_EXCEPTION);
+        submitReleasingTask(200);
+        String result = f.get(30000, TimeUnit.MILLISECONDS);
+
+        assertEquals("success", result);
+    }
+
+    @Test(timeout = 120000)
+    @Repeat(10)
+    public void getWithTimeout_timesOut() throws Exception {
+        ICompletableFuture<String> f = submitAwaitingTaskNoCallbacks(NO_EXCEPTION);
+
+        expected.expect(TimeoutException.class);
+        f.get(1, TimeUnit.MILLISECONDS);
+    }
+
+    @Test
+    public void singleCancellation_beforeDone_succeeds() throws Exception {
+        ICompletableFuture<String> f = submitAwaitingTaskNoCallbacks(NO_EXCEPTION);
+        assertTaskInExecution();
+
+        boolean cancelResult = f.cancel(false);
+
+        assertTrue("Task cancellation succeeded should succeed", cancelResult);
+    }
+
+    @Test
+    public void doubleCancellation_beforeDone_firstSucceeds_secondFails() throws Exception {
+        ICompletableFuture<String> f = submitAwaitingTaskNoCallbacks(NO_EXCEPTION);
+        assertTaskInExecution(); // but never released to execute logic
+
+        boolean firstCancelResult = f.cancel(false);
+        boolean secondCancelResult = f.cancel(false);
+
+        assertTrue("First task cancellation should succeed", firstCancelResult);
+        assertFalse("Second task cancellation should failed", secondCancelResult);
+    }
+
+    @Test
+    public void cancellation_afterDone_taskNotCancelled_flagsSetCorrectly() throws Exception {
+        ICompletableFuture<String> f = submitAwaitingTaskNoCallbacks(NO_EXCEPTION);
+        assertTaskInExecution();
+        releaseAwaitingTask();
+        assertTaskExecutedItsLogic();
+        assertTaskFinishedEventually();
+
+        boolean firstCancelResult = f.cancel(false);
+        boolean secondCancelResult = f.cancel(false);
+
+        assertFalse("Cancellation should not succeed after task is done", firstCancelResult);
+        assertFalse("Cancellation should not succeed after task is done", secondCancelResult);
+        assertTrue("Task should be done", f.isDone());
+        assertFalse("Task should NOT be cancelled", f.isCancelled());
+        assertEquals("success", f.get());
+    }
+
+    @Test
+    public void noCancellation_afterDone_flagsSetCorrectly() throws Exception {
+        ICompletableFuture<String> f = submitAwaitingTaskNoCallbacks(NO_EXCEPTION);
+        assertTaskInExecution();
+        releaseAwaitingTask();
+        assertTaskExecutedItsLogic();
+        assertTaskFinishedEventually();
+
+        assertTrue("Task should be done", f.isDone());
+        assertFalse("Task should NOT be cancelled", f.isCancelled());
+        assertEquals("success", f.get());
+    }
+
+
+    @Test(timeout = 60000)
+    @Repeat(10)
+    public void cancelAndGet_taskCancelled_withoutInterruption_logicExecuted() throws Exception {
+        ICompletableFuture<String> f = submitAwaitingTaskNoCallbacks(NO_EXCEPTION);
+        assertTaskInExecution();
+
+        boolean cancelResult = f.cancel(false);
+        releaseAwaitingTask();
+        assertTaskExecutedItsLogic(); // cancellation came, when task already awaiting, so logic executed
+        assertTaskFinishedEventually();
+
+        assertTrue("Task cancellation should succeed", cancelResult);
+        assertTrue("Task should be done", f.isDone());
+        assertTrue("Task should be cancelled", f.isCancelled());
+        expected.expect(CancellationException.class);
+        f.get();
+    }
+
+    @Test(timeout = 120000)
+    @Repeat(10)
+    public void cancelAndGet_taskCancelled_withInterruption_noLogicExecuted() throws Exception {
+        ICompletableFuture<String> f = submitAwaitingTaskNoCallbacks(NO_EXCEPTION);
+        assertTaskInExecution();
+
+        boolean cancelResult = f.cancel(true);
+
+        assertTaskInterruptedAndDidNotExecuteItsLogic();
+        assertTaskFinishedEventually(); // task did not have to be releases - interruption was enough
+        assertTrue("Task cancellation should succeed", cancelResult);
+        assertTrue("Task should be done", f.isDone());
+        assertTrue("Task should be cancelled", f.isCancelled());
+        expected.expect(CancellationException.class);
+        f.get();
+    }
+
+
+    private static void assertTestExceptionThrown(AtomicReference<?>... refs) {
         for (AtomicReference<?> ref : refs)
             assertThat("ExecutionException expected", ref.get(), instanceOf(ExecutionException.class));
         for (AtomicReference<?> ref : refs)
             assertThat("TEST_EXCEPTION expected as cause", ((Throwable) ref.get()).getCause(),
-                Matchers.<Throwable>sameInstance(TEST_EXCEPTION));
+                    Matchers.<Throwable>sameInstance(THROW_TEST_EXCEPTION));
     }
 
-    private ICompletableFuture<String> submit(final Runnable... rs) {
-        return executionService.asCompletableFuture(executionService.submit("default", new Callable<String>() {
+    private ICompletableFuture<String> submitAwaitingTaskNoCallbacks(final Exception exception) {
+        return submitAwaitingTask(0, exception);
+    }
+
+    private ICompletableFuture<String> submitAwaitingTask(Integer numberOfCallbacks, final Exception exception) {
+        callbacksDoneLatch = new CountDownLatch(numberOfCallbacks);
+        return submit(new Callable<String>() {
             @Override
-            public String call() {
-                for (Runnable r : rs) r.run();
-                return "success";
+            public String call() throws Exception {
+                inExecutionLatch.countDown();
+                try {
+                    assertOpenEventually(startLogicLatch);
+                    executedLogic.countDown();
+                    if (exception != null) throw exception;
+                    return "success";
+                } finally {
+                    finishedLatch.countDown();
+                }
             }
-        }));
+        });
     }
 
-    private Runnable awaitStartLatch() {
-        return new Runnable() { @Override public void run() {
-            assertOpenEventually(startLatch);
-        }};
+    private void submitReleasingTask(final long millisToAwaitBeforeRelease) {
+        submit(new Runnable() {
+            @Override
+            public void run() {
+                sleepAtLeastMillis(millisToAwaitBeforeRelease);
+                releaseAwaitingTask();
+            }
+        });
     }
 
-    private Runnable openStartLatch() {
-        return new Runnable() { @Override public void run() {
-            startLatch.countDown();
-        }};
+    private ICompletableFuture<String> submit(final Callable<String> callable) {
+        return executionService.asCompletableFuture(executionService.submit("default", callable));
     }
 
-    private static Runnable throwException() {
-        return new Runnable() { @Override public void run() {
-            throw TEST_EXCEPTION;
-        }};
+    private void submit(final Runnable runnable) {
+        executionService.submit("default", runnable);
     }
 
-    private ExecutionCallback<String> setRefAndBumpDoneLatch(final AtomicReference<Object> ref) {
+    private Integer expectedNumberOfCallbacks(int number) {
+        return number;
+    }
+
+    private void releaseAwaitingTask() {
+        startLogicLatch.countDown();
+    }
+
+    private void assertCallbacksExecutedEventually() {
+        assertOpenEventually(callbacksDoneLatch);
+    }
+
+    private void assertTaskExecutedItsLogic() {
+        assertOpenEventually(executedLogic);
+    }
+
+    private void assertTaskInterruptedAndDidNotExecuteItsLogic() {
+        assertEquals(1, executedLogic.getCount());
+    }
+
+    private void assertTaskFinishedEventually() {
+        assertOpenEventually(finishedLatch);
+    }
+
+    private void assertTaskInExecution() {
+        assertOpenEventually(inExecutionLatch);
+    }
+
+    private ExecutionCallback<String> storeTaskResponseToReference(final AtomicReference<Object> ref) {
         return new ExecutionCallback<String>() {
-            @Override public void onResponse(String response) {
+            @Override
+            public void onResponse(String response) {
                 doit(response);
             }
-            @Override public void onFailure(Throwable t) {
+
+            @Override
+            public void onFailure(Throwable t) {
                 doit(t);
             }
+
             private void doit(Object response) {
                 ref.set(response);
-                doneLatch.countDown();
+                callbacksDoneLatch.countDown();
             }
         };
     }

--- a/hazelcast/src/test/java/com/hazelcast/spi/impl/AbstractCompletableFutureTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/spi/impl/AbstractCompletableFutureTest.java
@@ -10,19 +10,20 @@ import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.HazelcastTestSupport;
 import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
-import com.hazelcast.util.ExceptionUtil;
+import com.hazelcast.test.annotation.Repeat;
 import org.junit.Before;
+import org.junit.Rule;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
+import org.junit.rules.ExpectedException;
 import org.junit.runner.RunWith;
 
-import java.util.concurrent.ExecutionException;
+import java.util.concurrent.CancellationException;
 import java.util.concurrent.Executor;
 import java.util.concurrent.Executors;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
 
-import static com.hazelcast.spi.impl.AbstractCompletableFuture.ExecutionCallbackNode;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertNull;
@@ -36,12 +37,18 @@ import static org.mockito.Mockito.verifyZeroInteractions;
 @Category({QuickTest.class, ParallelTest.class})
 public class AbstractCompletableFutureTest extends HazelcastTestSupport {
 
-    private final static Object FAKE_GET_RESPONSE = "foobar";
+    private final static Object RESULT = "foobar";
+
+    private final static String EXCEPTION_MESSAGE = "You screwed buddy!";
+    private final static Exception EXCEPTION = new RuntimeException(EXCEPTION_MESSAGE);
 
     private HazelcastInstance hz;
     private ILogger logger;
     private NodeEngineImpl nodeEngine;
     private Executor executor;
+
+    @Rule
+    public ExpectedException expected = ExpectedException.none();
 
     @Before
     public void setup() {
@@ -51,16 +58,268 @@ public class AbstractCompletableFutureTest extends HazelcastTestSupport {
         executor = Executors.newFixedThreadPool(1);
     }
 
-    // ==================== setResult ===========================================
-
     @Test
-    public void setResult_whenPendingCallback() {
-        setResult_whenPendingCallback(null);
-        setResult_whenPendingCallback("foo");
-        setResult_whenPendingCallback(new Exception());
+    public void future_notExecuted_notDoneNotCancelled() throws Exception {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+
+        assertFalse("New future should not be done", future.isDone());
+        assertFalse("New future should not be cancelled", future.isCancelled());
     }
 
-    public void setResult_whenPendingCallback(final Object result) {
+    @Test
+    public void future_notExecuted_callbackRegistered_notDoneNotCancelled() throws Exception {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+        future.andThen(mock(ExecutionCallback.class));
+
+        assertFalse("New future should not be done", future.isDone());
+        assertFalse("New future should not be cancelled", future.isCancelled());
+    }
+
+    @Test
+    public void future_ordinaryResultSet_doneNotCancelled() throws Exception {
+        future_resultSet_doneNotCancelled(RESULT);
+    }
+
+    @Test
+    public void future_nullResultSet_doneNotCancelled() throws Exception {
+        future_resultSet_doneNotCancelled(null);
+    }
+
+    @Test
+    public void future_exceptionResultSet_doneNotCancelled() throws Exception {
+        future_resultSet_doneNotCancelled(EXCEPTION);
+    }
+
+    private void future_resultSet_doneNotCancelled(Object result) {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+        future.setResult(result);
+
+        assertTrue("Future with result should be done", future.isDone());
+        assertFalse("Done future should not be cancelled", future.isCancelled());
+    }
+
+    @Test
+    public void future_resultNotSet_cancelled() throws Exception {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+        boolean cancelled = future.cancel(false);
+
+        assertTrue(cancelled);
+        assertTrue("Cancelled future should be done", future.isDone());
+        assertTrue("Cancelled future should be cancelled", future.isCancelled());
+    }
+
+    @Test
+    public void future_resultSetAndCancelled() throws Exception {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+        future.setResult(RESULT);
+        boolean cancelled = future.cancel(false);
+
+        assertFalse(cancelled);
+        assertTrue("Done future should be done", future.isDone());
+        assertFalse("Done future should not be cancelled even if cancelled executed", future.isCancelled());
+    }
+
+    @Test
+    public void future_cancelledAndResultSet() throws Exception {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+        boolean cancelled = future.cancel(false);
+        future.setResult(RESULT);
+
+        assertTrue(cancelled);
+        assertTrue("Cancelled future should be done", future.isDone());
+        assertTrue("Cancelled future should be cancelled", future.isCancelled());
+        assertNull("Internal result should be null", future.getResult());
+    }
+
+    @Test(expected = CancellationException.class)
+    public void get_cancelledFuture_exceptionThrown() throws Exception {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+
+        future.cancel(false);
+
+        future.get();
+    }
+
+    @Test(expected = CancellationException.class)
+    public void getWithTimeout_cancelledFuture_exceptionThrown() throws Exception {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+
+        future.cancel(false);
+
+        future.get(10, TimeUnit.MILLISECONDS);
+    }
+
+    @Test
+    public void get_ordinaryResultSet_returnsResult() throws Exception {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+        future.setResult(RESULT);
+
+        Object result = future.get();
+
+        assertSame(RESULT, result);
+    }
+
+    @Test
+    public void getWithTimeout_ordinaryResultSet_returnsResult() throws Exception {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+        future.setResult(RESULT);
+
+        Object result = future.get(10, TimeUnit.MILLISECONDS);
+
+        assertSame(RESULT, result);
+    }
+
+    @Test
+    public void get_exceptionResultSet_exceptionThrown() throws Exception {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+        future.setResult(EXCEPTION);
+
+        expected.expect(EXCEPTION.getClass());
+        expected.expectMessage(EXCEPTION_MESSAGE);
+
+        future.get();
+    }
+
+    @Test
+    public void getWithTimeout_exceptionResultSet_exceptionThrown_noTimeout() throws Exception {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+        future.setResult(EXCEPTION);
+
+        expected.expect(EXCEPTION.getClass());
+        expected.expectMessage(EXCEPTION_MESSAGE);
+
+        future.get(1, TimeUnit.NANOSECONDS);
+    }
+
+    @Test
+    public void get_nullResultSet_returnsResult() throws Exception {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+        future.setResult(null);
+
+        Object result = future.get();
+
+        assertNull(result);
+    }
+
+    @Test
+    public void getWithTimeout_nullResultSet_returnsResult() throws Exception {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+        future.setResult(null);
+
+        Object result = future.get(10, TimeUnit.MILLISECONDS);
+
+        assertNull(result);
+    }
+
+    @Test(expected = TimeoutException.class, timeout = 120000)
+    @Repeat(10)
+    public void getWithTimeout_resultNotSet_timesOut() throws Exception {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+
+        future.get(10, TimeUnit.MILLISECONDS);
+    }
+
+    @Test(expected = TimeoutException.class, timeout = 60000)
+    public void getWithTimeout_zeroTimeout_resultNotSet_timesOut() throws Exception {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+
+        future.get(0, TimeUnit.MILLISECONDS);
+    }
+
+    @Test(expected = TimeoutException.class, timeout = 60000)
+    public void getWithTimeout_negativeTimeout_resultNotSet_timesOut() throws Exception {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+
+        future.get(-1, TimeUnit.MILLISECONDS);
+    }
+
+    @Test(expected = TimeoutException.class, timeout = 60000)
+    public void getWithTimeout_lowerThanOneMilliTimeout_resultNotSet_timesOut() throws Exception {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+
+        future.get(1, TimeUnit.NANOSECONDS);
+    }
+
+    @Test
+    public void getWithTimeout_threadInterrupted_exceptionThrown() throws Exception {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+
+        Thread.currentThread().interrupt();
+
+        expected.expect(InterruptedException.class);
+        future.get(100, TimeUnit.MILLISECONDS);
+    }
+
+    @Test(timeout = 60000)
+    public void getWithTimeout_waited_notifiedOnSet() throws Exception {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+
+        submitSetResultAfterTimeInMillis(future, RESULT, 200);
+        Object result = future.get(30000, TimeUnit.MILLISECONDS);
+
+        assertEquals(RESULT, result);
+    }
+
+    @Test(timeout = 60000)
+    public void getWithTimeout_waited_waited_notifiedOnCancel() throws Exception {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+
+        submitCancelAfterTimeInMillis(future, 200);
+
+        expected.expect(CancellationException.class);
+        future.get(30000, TimeUnit.MILLISECONDS);
+    }
+
+    @Test
+    public void setResult_ordinaryResultSet_futureDone() throws Exception {
+        setResult_resultSet_futureDone(RESULT);
+    }
+
+    @Test
+    public void setResult_exceptionResultSet_futureDone() throws Exception {
+        setResult_resultSet_futureDone(EXCEPTION);
+    }
+
+    @Test
+    public void setResult_nullResultSet_futureDone() throws Exception {
+        setResult_resultSet_futureDone(null);
+    }
+
+    private void setResult_resultSet_futureDone(Object result) {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+
+        future.setResult(result);
+
+        assertTrue("Future should be done after result has been set", future.isDone());
+    }
+
+    @Test
+    public void setResult_whenResultAlreadySet_secondResultDiscarded() throws Exception {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+        Object initialResult = "firstresult", secondResult = "secondresult";
+
+        future.setResult(initialResult);
+        future.setResult(secondResult);
+
+        assertSame(initialResult, future.get());
+    }
+
+    @Test
+    public void setResult_whenPendingCallback_nullResult() {
+        setResult_whenPendingCallback_callbacksExecutedCorrectly(null);
+    }
+
+    @Test
+    public void setResult_whenPendingCallback_ordinaryResult() {
+        setResult_whenPendingCallback_callbacksExecutedCorrectly("foo");
+    }
+
+    @Test
+    public void setResult_whenPendingCallback_exceptionResult() {
+        setResult_whenPendingCallback_callbacksExecutedCorrectly(new Exception());
+    }
+
+    public void setResult_whenPendingCallback_callbacksExecutedCorrectly(final Object result) {
         FutureImpl future = new FutureImpl(nodeEngine, logger);
         final ExecutionCallback callback1 = mock(ExecutionCallback.class);
         final ExecutionCallback callback2 = mock(ExecutionCallback.class);
@@ -69,7 +328,6 @@ public class AbstractCompletableFutureTest extends HazelcastTestSupport {
 
         future.setResult(result);
 
-        assertSame(result, future.state);
         assertTrueEventually(new AssertTask() {
             @Override
             public void run() throws Exception {
@@ -80,7 +338,6 @@ public class AbstractCompletableFutureTest extends HazelcastTestSupport {
                 }
             }
         });
-
         assertTrueEventually(new AssertTask() {
             @Override
             public void run() throws Exception {
@@ -94,120 +351,105 @@ public class AbstractCompletableFutureTest extends HazelcastTestSupport {
     }
 
     @Test
-    public void setResult_whenNoPendingCallback() {
-        setResult_whenNoPendingCallback(null);
-        setResult_whenNoPendingCallback("foo");
-    }
-
-    public void setResult_whenNoPendingCallback(Object result) {
-        FutureImpl future = new FutureImpl(nodeEngine, logger);
-
-        future.setResult(result);
-
-        assertSame(result, future.state);
-    }
-
-
-    @Test
-    public void setResult_whenResultAlreadySet() {
-        FutureImpl future = new FutureImpl(nodeEngine, logger);
-        Object initialResult = "firstresult";
-
-        future.setResult(initialResult);
-
-        future.setResult("secondresult");
-        assertSame(initialResult, future.state);
-    }
-
-    // ==================== get ===========================================
-
-    @Test
-    public void get() throws ExecutionException, InterruptedException {
-        FutureImpl future = new FutureImpl(nodeEngine, logger);
-
-        Object result = future.get();
-
-        assertSame(FAKE_GET_RESPONSE, result);
-        assertEquals(Long.MAX_VALUE, future.timeout);
-        assertEquals(TimeUnit.MILLISECONDS, future.unit);
-    }
-
-    @Test
-    public void get_whenTimeout() throws ExecutionException, InterruptedException {
-        FutureImpl future = new FutureImpl(nodeEngine, logger);
-        future.getException = new TimeoutException();
-        Object result = future.get();
-
-        assertNull(result);
-    }
-
-    // ==================== getResult ===========================================
-
-    @Test
-    public void getResult_whenInitialState() {
+    public void getResult_whenInitialState() throws Exception {
         FutureImpl future = new FutureImpl(nodeEngine, logger);
 
         Object result = future.getResult();
 
-        assertNull(result);
+        assertNull("Internal result should be null initially", result);
     }
 
     @Test
-    public void getResult_whenPendingCallback() {
+    public void getResult_whenPendingCallback() throws Exception {
         FutureImpl future = new FutureImpl(nodeEngine, logger);
         future.andThen(mock(ExecutionCallback.class));
 
         Object result = future.getResult();
 
-        assertNull(result);
+        assertNull("Internal result should be null initially", result);
     }
 
     @Test
-    public void getResult_whenNullResult() {
+    public void getResult_whenNullResult() throws Exception {
         FutureImpl future = new FutureImpl(nodeEngine, logger);
         future.setResult(null);
 
         Object result = future.getResult();
 
-        assertNull(result);
+        assertNull("Internal result should be null when set to null", result);
     }
 
-    @Test
-    public void getResult_whenNoneNullResult() {
-        FutureImpl future = new FutureImpl(nodeEngine, logger);
-        Object expectedResult = "result";
-        future.setResult(expectedResult);
-
-        Object result = future.getResult();
-
-        assertSame(expectedResult, result);
-    }
-
-    // ==================== andThen ===========================================
 
     @Test(expected = IllegalArgumentException.class)
-    public void andThen_whenNullCallback() {
+    public void andThen_whenNullCallback_exceptionThrown() {
         FutureImpl future = new FutureImpl(nodeEngine, logger);
         future.andThen(null, executor);
     }
 
     @Test(expected = IllegalArgumentException.class)
-    public void andThen_whenNullExecutor() {
+    public void andThen_whenNullExecutor_exceptionThrown() {
         FutureImpl future = new FutureImpl(nodeEngine, logger);
         future.andThen(mock(ExecutionCallback.class), null);
     }
 
     @Test
-    public void andThen_whenResultAvailable() {
+    public void andThen_whenInitialState() {
         FutureImpl future = new FutureImpl(nodeEngine, logger);
-        final Object result = "result";
-        future.setResult(result);
+        ExecutionCallback callback = mock(ExecutionCallback.class);
 
-        final ExecutionCallback callback = mock(ExecutionCallback.class);
         future.andThen(callback, executor);
 
-        assertSame(result, future.state);
+        verifyZeroInteractions(callback);
+    }
 
+    @Test
+    public void andThen_whenCancelled() {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+        ExecutionCallback callback = mock(ExecutionCallback.class);
+
+        future.cancel(false);
+        future.andThen(callback, executor);
+
+        verifyZeroInteractions(callback);
+    }
+
+    @Test
+    public void andThen_whenPendingCallback() {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+        ExecutionCallback callback1 = mock(ExecutionCallback.class);
+        ExecutionCallback callback2 = mock(ExecutionCallback.class);
+
+        future.andThen(callback1, executor);
+        future.andThen(callback2, executor);
+
+        verifyZeroInteractions(callback1);
+        verifyZeroInteractions(callback2);
+    }
+
+    @Test
+    public void andThen_whenPendingCallback_andCancelled() {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+        ExecutionCallback callback1 = mock(ExecutionCallback.class);
+        ExecutionCallback callback2 = mock(ExecutionCallback.class);
+
+        future.andThen(callback1, executor);
+        future.cancel(false);
+        future.andThen(callback2, executor);
+
+        verifyZeroInteractions(callback1);
+        verifyZeroInteractions(callback2);
+    }
+
+    @Test
+    public void andThen_whenResultAvailable() throws Exception {
+        FutureImpl future = new FutureImpl(nodeEngine, logger);
+        final Object result = "result";
+        final ExecutionCallback callback = mock(ExecutionCallback.class);
+
+        future.setResult(result);
+        future.andThen(callback, executor);
+
+        assertSame(result, future.get());
         assertTrueEventually(new AssertTask() {
             @Override
             public void run() throws Exception {
@@ -216,112 +458,39 @@ public class AbstractCompletableFutureTest extends HazelcastTestSupport {
         });
     }
 
-    @Test
-    public void andThen_whenInitialState() {
-        FutureImpl future = new FutureImpl(nodeEngine, logger);
-
-        ExecutionCallback callback = mock(ExecutionCallback.class);
-        future.andThen(callback, executor);
-
-        ExecutionCallbackNode node = assertInstanceOf(ExecutionCallbackNode.class, future.state);
-        assertSame(node.callback, callback);
-        assertSame(node.executor, executor);
-        assertSame(node.next, AbstractCompletableFuture.INITIAL_STATE);
-        verifyZeroInteractions(callback);
+    private void submitCancelAfterTimeInMillis(final FutureImpl future, final int timeInMillis) {
+        submit(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    sleepMillis(timeInMillis);
+                } finally {
+                    future.cancel(false);
+                }
+            }
+        });
     }
 
-    @Test
-    public void andThen_whenPendingCallback() {
-        FutureImpl future = new FutureImpl(nodeEngine, logger);
-
-        ExecutionCallback callback1 = mock(ExecutionCallback.class);
-        future.andThen(callback1, executor);
-
-        ExecutionCallback callback2 = mock(ExecutionCallback.class);
-        future.andThen(callback2, executor);
-
-        ExecutionCallbackNode secondNode = assertInstanceOf(ExecutionCallbackNode.class, future.state);
-        assertSame(secondNode.callback, callback2);
-        assertSame(secondNode.executor, executor);
-
-        ExecutionCallbackNode firstNode = secondNode.next;
-        assertSame(firstNode.callback, callback1);
-        assertSame(firstNode.executor, executor);
-        assertSame(firstNode.next, AbstractCompletableFuture.INITIAL_STATE);
-
-        verifyZeroInteractions(callback1);
-        verifyZeroInteractions(callback2);
+    private void submitSetResultAfterTimeInMillis(final FutureImpl future, final Object result, final int timeInMillis) {
+        submit(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    sleepMillis(timeInMillis);
+                } finally {
+                    future.setResult(result);
+                }
+            }
+        });
     }
 
-    // ==================== isDone ===========================================
-
-    @Test
-    public void isDone_whenInitialState() {
-        FutureImpl future = new FutureImpl(nodeEngine, logger);
-
-        boolean result = future.isDone();
-
-        assertFalse(result);
+    private void submit(Runnable runnable) {
+        new Thread(runnable).start();
     }
 
-    @Test
-    public void isDone_whenCallbackRegistered() {
-        FutureImpl future = new FutureImpl(nodeEngine, logger);
-        future.andThen(mock(ExecutionCallback.class));
-
-        boolean result = future.isDone();
-
-        assertFalse(result);
-    }
-
-    @Test
-    public void isDone_whenNullResultSet() {
-        FutureImpl future = new FutureImpl(nodeEngine, logger);
-        future.setResult(null);
-
-        boolean result = future.isDone();
-
-        assertTrue(result);
-    }
-
-    @Test
-    public void isDone_whenNoneNullResultSet() {
-        FutureImpl future = new FutureImpl(nodeEngine, logger);
-        future.setResult("foo");
-
-        boolean result = future.isDone();
-
-        assertTrue(result);
-    }
-
-    class FutureImpl extends AbstractCompletableFuture {
-
-        private long timeout;
-        private TimeUnit unit;
-        private Exception getException;
-
+    private class FutureImpl extends AbstractCompletableFuture<Object> {
         protected FutureImpl(NodeEngine nodeEngine, ILogger logger) {
             super(nodeEngine, logger);
-        }
-
-        @Override
-        public boolean cancel(boolean mayInterruptIfRunning) {
-            return false;
-        }
-
-        @Override
-        public boolean isCancelled() {
-            return false;
-        }
-
-        @Override
-        public Object get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
-            this.timeout = timeout;
-            this.unit = unit;
-            if(getException!=null){
-                ExceptionUtil.sneakyThrow(getException);
-            }
-            return FAKE_GET_RESPONSE;
         }
     }
 }

